bigchaindb/elections/election.py

```diff
@@ -242,24 +244,41 @@ def show_election(self, bigchain):
         return response
 
     @classmethod
-    def approved_update(cls, bigchain, new_height, txns):
-        votes = {}
-        for txn in txns:
-            if not isinstance(txn, Vote):
+    def approved_elections(cls, bigchain, new_height, txns):
+        elections = defaultdict(list)
+        for tx in txns:
+            if not isinstance(tx, Vote):
                 continue
+            election_id = tx.asset['id']
+            elections[election_id].append(tx)
 
-            election_id = txn.asset['id']
-            election_votes = votes.get(election_id, [])
-            election_votes.append(txn)
-            votes[election_id] = election_votes
-
-            election = cls.has_concluded(bigchain, election_id, election_votes, new_height)
-            # Once an election concludes any other conclusion for the same
-            # or any other election is invalidated
-            if election:
-                cls.store_election_results(bigchain, election, new_height)
-                return cls.on_approval(bigchain, election, new_height)
-        return None
+        validator_set_updated = False
+        validator_set_change = []
+        for election_id, votes in elections.items():
+            election = bigchain.get_transaction(election_id)
+
+            if not election.has_concluded(bigchain, election_id, votes, new_height):
+                continue
+
+            if election.makes_validator_set_change():
+                if validator_set_updated:
+                    continue
+                validator_set_change.append(election.get_validator_set_change(bigchain, new_height))
+                validator_set_updated = True
+
+            election.on_approval(bigchain, election, new_height)
+
+        return validator_set_change
+
+    def makes_validator_set_change(self):
+        return self.CHANGES_VALIDATOR_SET
+
+    def get_validator_set_change(self, bigchain, new_height):
+        if self.makes_validator_set_change():
```

Comment:
This `if` seems unnecessary i.e. looking at the logic from `approved_elections` this function will only be called if `makes_validator_set_change()` returns `True`?

Reply:
The idea is that `get_validator_set_change` is a generic function, that is safe to call on any approved election. If the election has `CHANGES_VALIDATOR_SET == False`, then nothing happens. If the election has `CHANGES_VALIDATOR_SET == True`, then we execute the custom code to change the validator set.

The reason I built it this way is so that I can set the base `Election` class with `CHANGES_VALIDATOR_SET == True` and have the `change_validator_set` function raise a `NotImplementedError`. This addressed the concern Lev raised that someone could extend the base class and not understand how validator updates work. Anyone who makes a custom election class needs to either explicitly set `CHANGES_VALIDATOR_SET` to `False`, or deal with implementing the `change_validator_set` function, so in either case, they can't ignore the validator set issue.

Reply:
@z-bowen such an interface might be helpful indeed but one does not expect the function named `get_validator_set_change` to update the database. What about doing the same before calling `on_approval`?

```
if self.CHANGES_VALIDATOR_SET:
    self.change_validator_set()
election.on_approval(...)
```

Reply:
I agree but it would be much easy to understand if the change to the validator set or any state mutation is done in `on_approval` method and nowhere else.

Reply:
Yeah, that's a tradeoff. I am fine with doing everything in `on_approval` too.